bundles/core/org.openhab.core.compat1x/src/main/java/org/openhab/core/types/TypeParser.java

```diff
@@ -34,15 +34,27 @@
 	 * @return the corresponding State instance or <code>null</code>
 	 */
 	public static State parseState(List<Class<? extends State>> types, String s) {
-		for(Class<? extends Type> type : types) {
+		for (Class<? extends Type> type : types) {
 			try {									
 				Method valueOf = type.getMethod("valueOf", String.class);
 				State state = (State) valueOf.invoke(type, s);
-				if(state!=null) return state;
+				if (state != null) {
+					return state;
+				}
 			} catch (NoSuchMethodException e) {
 			} catch (IllegalArgumentException e) {
 			} catch (IllegalAccessException e) {
 			} catch (InvocationTargetException e) {
+				// we need this special treatment because UnDefType has
+				// overridden its toString() method and hence valueOf does not 
+				// work anymore (for the values being returned by toString())
+				if (type.equals(org.openhab.core.types.UnDefType.class)) {
+					if (UnDefType.UNDEF.toString().equals(s)) {
+						return UnDefType.UNDEF;
+					} else if (UnDefType.NULL.toString().equals(s)) {
+						return UnDefType.NULL;
+					}
```

Comment:
If this has to be done outside of `UnDefType`, I'd add an else { assert... } here since you can't tell if some new ENUM state ever gets added to `UnDefType`.

I'd also add an assert to the legacy ...Delegate handler in the case where you pass it a non-null value, and the `TypeParser.parseState` call returns null.  Currently this falls through, and passes null to the ESH layer, only to let it fail, so it takes backtracking to work out why...